src/main/kotlin/com/coder/gateway/CoderGatewayConnectionProvider.kt

```diff
@@ -2,93 +2,129 @@
+
+            val cli = CoderCLIManager.ensureCLI(
+                deploymentURL.toURL(),
+                client.buildInfo().version,
+                settings,
+                indicator,
+            )
+
+            indicator.text = "Authenticating Coder CLI..."
+            cli.login(client.token)
+
+            indicator.text = "Configuring Coder CLI..."
+            cli.configSsh(agents)
+
+            // TODO: Ask for these if missing.  Maybe we can reuse the second
+            //  step of the wizard?  Could also be nice if we automatically used
+            //  the last IDE.
+            if (parameters[IDE_PRODUCT_CODE].isNullOrBlank()) {
+                throw IllegalArgumentException("Query parameter \"$IDE_PRODUCT_CODE\" is missing")
+            }
+            if (parameters[IDE_BUILD_NUMBER].isNullOrBlank()) {
+                throw IllegalArgumentException("Query parameter \"$IDE_BUILD_NUMBER\" is missing")
+            }
+            if (parameters[IDE_PATH_ON_HOST].isNullOrBlank() && parameters[IDE_DOWNLOAD_LINK].isNullOrBlank()) {
+                throw IllegalArgumentException("One of \"$IDE_PATH_ON_HOST\" or \"$IDE_DOWNLOAD_LINK\" is required")
+            }
+
+            // TODO: Ask for the project path if missing and validate the path.
+            val folder = parameters[FOLDER] ?: throw IllegalArgumentException("Query parameter \"$FOLDER\" is missing")
+
+            parameters
+                .withWorkspaceHostname(CoderCLIManager.getHostName(deploymentURL.toURL(), workspace))
+                .withProjectPath(folder)
+                .withWebTerminalLink(client.url.withPath("/@$username/$workspace.name/terminal").toString())
+                .withConfigDirectory(cli.coderConfigPath.toString())
+                .withName(name)
+        }
         return null
     }
 
+    /**
+     * Return an authenticated Coder CLI and the user's name, asking for the
+     * token as long as it continues to result in an authentication failure.
+     */
+    private fun authenticate(deploymentURL: URL, queryToken: String?, lastToken: Pair<String, TokenSource>? = null): Pair<CoderRestClient, String> {
+        // Use the token from the query, unless we already tried that.
+        val isRetry = lastToken != null
+        val token = if (!queryToken.isNullOrBlank() && !isRetry)
+            Pair(queryToken, TokenSource.QUERY)
+        else CoderRemoteConnectionHandle.askToken(
+            deploymentURL,
+            lastToken,
+            isRetry,
+            useExisting = true,
+        )
+        if (token == null) { // User aborted.
+            throw IllegalArgumentException("Unable to connect to $deploymentURL, $TOKEN is missing")
+        }
+        val client = CoderRestClient(deploymentURL, token.first)
+        return try {
+            Pair(client, client.me().username)
+        } catch (ex: AuthenticationResponseException) {
+            authenticate(deploymentURL, queryToken, token)
```

Comment:
There is a limit to how deep this can go; I don't think we'll ever hit it.
But it's probably simpler to just use a loop instead.

Reply:
I think we avoid overflow because Kotlin would do tail recursion here, although I would not be surprised if I understood it incorrectly.  :laughing: